src/SignalR/common/Protocols.Json/src/Protocol/JsonHubProtocol.cs

```diff
@@ -0,0 +1,758 @@
+                        }
+
+                        headers[propertyName] = reader.GetString();
+                        break;
+                    case JsonTokenType.Comment:
+                        break;
+                    case JsonTokenType.EndObject:
+                        return headers;
+                }
+            }
+
+            throw new JsonReaderException("Unexpected end when reading message headers", 0, 0);
+        }
+
+        private void WriteMessageCore(HubMessage message, IBufferWriter<byte> stream)
+        {
+            var writer = new Utf8JsonWriter(stream);
+
+            writer.WriteStartObject();
+            switch (message)
+            {
+                case InvocationMessage m:
+                    WriteMessageType(ref writer, HubProtocolConstants.InvocationMessageType);
+                    WriteHeaders(ref writer, m);
+                    WriteInvocationMessage(m, ref writer);
+                    break;
+                case StreamInvocationMessage m:
+                    WriteMessageType(ref writer, HubProtocolConstants.StreamInvocationMessageType);
+                    WriteHeaders(ref writer, m);
+                    WriteStreamInvocationMessage(m, ref writer);
+                    break;
+                case StreamItemMessage m:
+                    WriteMessageType(ref writer, HubProtocolConstants.StreamItemMessageType);
+                    WriteHeaders(ref writer, m);
+                    WriteStreamItemMessage(m, ref writer);
+                    break;
+                case CompletionMessage m:
+                    WriteMessageType(ref writer, HubProtocolConstants.CompletionMessageType);
+                    WriteHeaders(ref writer, m);
+                    WriteCompletionMessage(m, ref writer);
+                    break;
+                case CancelInvocationMessage m:
+                    WriteMessageType(ref writer, HubProtocolConstants.CancelInvocationMessageType);
+                    WriteHeaders(ref writer, m);
+                    WriteCancelInvocationMessage(m, ref writer);
+                    break;
+                case PingMessage _:
+                    WriteMessageType(ref writer, HubProtocolConstants.PingMessageType);
+                    break;
+                case CloseMessage m:
+                    WriteMessageType(ref writer, HubProtocolConstants.CloseMessageType);
+                    WriteCloseMessage(m, ref writer);
+                    break;
+                default:
+                    throw new InvalidOperationException($"Unsupported message type: {message.GetType().FullName}");
+            }
+            writer.WriteEndObject();
+            writer.Flush();
+        }
+
+        private void WriteHeaders(ref Utf8JsonWriter writer, HubInvocationMessage message)
+        {
+            if (message.Headers != null && message.Headers.Count > 0)
+            {
+                writer.WriteStartObject(HeadersPropertyNameBytes);
```

Comment:
When writing a known property name, add `escape: false`. This will tell the writer to skip inspecting the property name and escaping characters. We don't need it because we know this is a valid property name.

It provides a small performance boost.